File: wizard.lua
#!/usr/bin/env lua

local cursor_theme = {
	name = "dragon_byte"
}

local function print_title(
	left_indentation,
	text
)
	print(
		left_indentation ..
		"\x1b[1;32m" ..
		text ..
		"\x1b[0m"
	)
	return
end

local function print_topic(
	left_indentation,
	text
)
	print(
		left_indentation ..
		"\x1b[32m*\x1b[0m " ..
		text
	)
	return
end

local function print_error(message)
	print(
		"\x1b[1;31mOpsie!\x1b[0m " ..
		message
	)
	return
end

local function highlight(text)
	return (
		"\x1b[31m" ..
		text ..
		"\x1b[0m"
	)
end

local function get_user()
	local whoami_stream = io.popen("whoami")
	local output = whoami_stream:read()
	whoami_stream:close()
	return (output)
end

local function get_installation_directory(user)
	return (
		user == "root" and
		"/usr/share/icons" or
		"/home/" ..
		user ..
		"/.local/share/icons"
	)
end

local function print_usage_instructions()
	print_title(
		"",
		"Usage Instructions"
	)
	print_title(
		"\t",
		"Starting Point"
	)
	print(
		"\t\tThis is a script to manage the build, install and uninstall of " ..
		cursor_theme.name ..
		"."
	)
	print("\t\tIt must be run from the root directory of the repository.")
	print_title(
		"\t",
		"Syntax"
	)
	print("\t\tUse it with the following syntax:")
	print(
		"\t\t\t" ..
		highlight("./wizard.lua") ..
		" <subcommand>"
	)
	print("\t\tThe subcommands it can accept are:")
	print(
		"\t\t\t" ..
		highlight("build") ..
		": builds the cursor theme and leave it available in your current directory."
	)
	print(
		"\t\t\t" ..
		highlight("install") ..
		": builds and installs the cursor for your user."
	)
	print("\t\t\tIf you are root, the installation will be at /usr/share/icons.")
	print("\t\t\tIf you are a normal user, the installation will be at ${HOME}/.local/share/icons.")
	print(
		"\t\t\t" ..
		highlight("uninstall") ..
		": uninstall the cursor for your user."
	)
	print("\t\t\tIt removes the installation directory used for an installation for your user.")
	return
end

local function create_directory_structure()
	print_topic(
		"\t",
		"Creating directory structure."
	)
	os.execute(
		"rm -rf ./" ..
		cursor_theme.name
	)
	os.execute(
		"mkdir -p ./" ..
		cursor_theme.name ..
		"/cursors"
	)
	return
end

local function create_metadata_files()
	print_topic(
		"\t",
		"Creating metadata files."
	)
	local index_file_stream = io.open(
		"./" ..
		cursor_theme.name ..
		"/index.theme",
		"w"
	)
	index_file_stream:write(
		"[Icon Theme]\nName=" ..
		cursor_theme.name ..
		"\n"
	)
	index_file_stream:close()
	return
end

local function create_cursor_files()
	print_topic(
		"\t",
		"Creating cursor files:"
	)
	local settings_directory_stream = io.popen("ls -A ./settings")
	for settings_file in settings_directory_stream:lines() do
		os.execute(
			"xcursorgen ./settings/" ..
			settings_file ..
			" > ./" ..
			cursor_theme.name ..
			"/cursors/" ..
			settings_file
		)
		print_topic(
			"\t\t",
			"Created cursor file " ..
			highlight(settings_file) ..
			"."
		)
	end
	settings_directory_stream:close()
	return
end

local function create_symbolic_links()
	print_topic(
		"\t",
		"Creating symbolic links:"
	)
	local symbolic_links = {
		{
			origin = "left_ptr",
			destination = {
				"pointing_hand",
				"pointer"
			}
		},
		{
			origin = "xterm",
			destination = {
				"text"
			}
		},
		{
			origin = "hand2",
			destination = {
				"hand",
				"hand1",
				"grab",
				"openhand"
			}
		},
		{
			origin = "help",
			destination = {
				"context-menu",
				"question_arrow",
				"dnd-ask"
			}
		},
		{
			origin = "link",
			destination = {
				"copy",
				"alias",
				"dnd-copy"
			}
		},
		{
			origin = "not-allowed",
			destination = {
				"no-drop",
				"dnd-no-drop",
				"dnd-none"
			}
		},
		{
			origin = "ns-resize",
			destination = {
				"top_side",
				"bottom_side",
				"n-resize",
				"s-resize",
				"split_v",
				"row-resize",
				"sb_v_double_arrow",
				"size_ver"
			}
		},
		{
			origin = "ew-resize",
			destination = {
				"left_side",
				"right_side",
				"e-resize",
				"w-resize",
				"split_h",
				"col-resize",
				"sb_h_double_arrow",
				"size_hor"
			}
		},
		{
			origin = "nesw-resize",
			destination = {
				"ne-resize",
				"sw-resize",
				"top_right_corner",
				"bottom_left_corner",
				"size_bdiag"
			}
		},
		{
			origin = "nwse-resize",
			destination = {
				"nw-resize",
				"se-resize",
				"top_left_corner",
				"bottom_right_corner",
				"size_fdiag"
			}
		},
		{
			origin = "all-scroll",
			destination = {
				"fleur",
				"cell",
				"move",
				"crosshair",
				"grabbing",
				"closedhand",
				"dnd-move",
				"cross",
				"plus",
				"size_all"
			}
		},
		{
			origin = "watch",
			destination = {
				"wait",
				"progress",
				"half-busy",
				"left_ptr_watch"
			}
		}
	}
	for symbolic_link_iterator =
		1,
		#symbolic_links
	do
		for destination_iterator =
			1,
			#symbolic_links[symbolic_link_iterator].destination
		do
			os.execute(
				"ln -sf " ..
				symbolic_links[symbolic_link_iterator].origin ..
				" " ..
				"./" ..
				cursor_theme.name ..
				"/cursors/" ..
				symbolic_links[symbolic_link_iterator].destination[destination_iterator]
			)
			print_topic(
				"\t\t",
				"Created symbolic link between " ..
				highlight(symbolic_links[symbolic_link_iterator].origin) ..
				" and " ..
				highlight(symbolic_links[symbolic_link_iterator].destination[destination_iterator]) ..
				"."
			)
		end
	end
	return
end

local function build_cursor_theme()
	print_title(
		"",
		"Building Cursor Theme"
	)
	create_directory_structure()
	create_metadata_files()
	create_cursor_files()
	create_symbolic_links()
	print(
		"\tCursor is available in the current directory as the directory " ..
		highlight(cursor_theme.name) ..
		"."
	)
	return
end

local function install_cursor_theme()
	build_cursor_theme()
	print_title(
		"",
		"Installing Cursor Theme"
	)
	local user = get_user()
	local installation_directory = get_installation_directory(user)
	os.execute(
		"mkdir -p " ..
		installation_directory
	)
	os.execute(
		"rm -rf " ..
		installation_directory ..
		"/" ..
		cursor_theme.name
	)
	os.execute(
		"mv ./" ..
		cursor_theme.name ..
		" " ..
		installation_directory ..
		"/" ..
		cursor_theme.name
	)
	print_topic(
		"\t",
		"Installed cursor theme at " ..
		highlight(
			installation_directory ..
			"/" ..
			cursor_theme.name
		) ..
		"."
	)
	print("\tThe cursor build has been moved to the installation directory.")
	return
end

local function uninstall_cursor_theme()
	print_title(
		"",
		"Uninstalling Cursor Theme"
	)
	local user = get_user()
	local installation_directory = get_installation_directory(user)
	os.execute(
		"rm -rf " ..
		installation_directory ..
		"/" ..
		cursor_theme.name
	)
	print_topic(
		"\t",
		"Uninstalled cursor theme at " ..
		highlight(
			installation_directory ..
			"/" ..
			cursor_theme.name
		) ..
		"."
	)
	return
end

local function main()
	if (arg[1] == "build") then
		build_cursor_theme()
	elseif (arg[1] == "install") then
		install_cursor_theme()
	elseif (arg[1] == "uninstall") then
		uninstall_cursor_theme()
	else
		print_usage_instructions()
	end
	return
end

main()

File: wizard.py
import os
import shutil

def get_current_directory() -> str:
	return os.path.dirname(__file__)

class SymlinkPair:
	def __init__(
		self,
		properties
	):
		self._source = properties["source"]
		self._outcomes = properties["outcomes"]

	def get_source(self) -> str:
		return self._source

	def get_outcomes(self) -> list[str]:
		return self._outcomes

class Cursor:
	def __init__(
		self,
		properties
	):
		self._name: str = properties["name"]
		self._output_directory: str = properties["output_directory"]
		self._settings_directory: str = properties["settings_directory"]
		self._symlink_pairs: list[SymlinkPair] = properties["symlink_pairs"]

	def get_name(self) -> str:
		return self._name

	def get_output_directory(self) -> str:
		return self._output_directory
	
	def get_settings_directory(self) -> str:
		return self._settings_directory
	
	def get_symlink_pairs(self) -> list[SymlinkPair]:
		return self._symlink_pairs

class CursorBuilder:
	def __init__(
		self,
		properties
	):
		self._cursor: Cursor = properties["cursor"]

	def _create_cursor_output_directory(self):
		shutil.rmtree(self._cursor.get_output_directory())
		os.mkdir(self._cursor.get_output_directory())
		os.mkdir(
			os.path.join(
				self._cursor.get_output_directory(),
				"cursors"
			)
		)

	def _create_index_file(self):
		path = os.path.join(
			self._cursor.get_output_directory(),
			"index.theme"
		)
		file = open(
			path,
			"w"
		)
		file.write(f"[Index Theme]\nName={self._cursor.get_name()}")
		file.close()

	def _get_settings_files(self) -> list[str]:
		return os.listdir(self._cursor.get_settings_directory())

	def _build_cursor_files(
		self,
		settings_files
	):
		for file in settings_files:
			source_path = os.path.join(
				self._cursor.get_settings_directory(),
				file
			)
			output_path = os.path.join(
				self._cursor.get_output_directory(),
				"cursors",
				file
			)
			os.system(f"xcursorgen {source_path} > {output_path}")

	def _create_symlink_pairs(self):
		for symlink_pair in self._cursor.get_symlink_pairs():
			for outcome in symlink_pair.get_outcomes():
				outcome_path = os.path.join(
					self._cursor.get_output_directory(),
					"cursors",
					outcome
				)
				os.system(f"ln -sf {symlink_pair.get_source()} {outcome_path}")

	def build(self):
		self._create_cursor_output_directory()
		self._create_index_file()
		self._build_cursor_files(self._get_settings_files())
		self._create_symlink_pairs()

def main():
	dragon_byte = Cursor({
		"name": "Dragon Byte",
		"output_directory": os.path.join(
			get_current_directory(),
			"dragon_byte"
		),
		"settings_directory": os.path.join(
			get_current_directory(),
			"settings"
		),
		"symlink_pairs": [
			SymlinkPair({
				"source": "left_ptr",
				"outcomes": [
					"pointing_hand",
					"pointer"
				]
			}),
			SymlinkPair({
				"source": "xterm",
				"outcomes": [ "text" ]
			}),
			SymlinkPair({
				"source": "hand2",
				"outcomes": [
					"hand",
					"hand1",
					"grab",
					"openhand"
				]
			}),
			SymlinkPair({
				"source": "help",
				"outcomes": [
					"context-menu",
					"question_arrow",
					"dnd-ask"
				]
			}),
			SymlinkPair({
				"source": "link",
				"outcomes": [
					"copy",
					"alias",
					"dnd-copy"
				]
			}),
			SymlinkPair({
				"source": "not-allowed",
				"outcomes": [
					"no-drop",
					"dnd-no-drop",
					"dnd-none"
				]
			}),
			SymlinkPair({
				"source": "ns-resize",
				"outcomes": [
					"top_side",
					"bottom_side",
					"n-resize",
					"s-resize",
					"split_v",
					"row-resize",
					"sb_v_double_arrow",
					"size_ver"
				]
			}),
			SymlinkPair({
				"source": "ew-resize",
				"outcomes": [
					"left_side",
					"right_side",
					"e-resize",
					"w-resize",
					"split_h",
					"col-resize",
					"sb_h_double_arrow",
					"size_hor"
				]
			}),
			SymlinkPair({
				"source": "nesw-resize",
				"outcomes": [
					"ne-resize",
					"sw-resize",
					"top_right_corner",
					"bottom_left_corner",
					"size_bdiag"
				]
			}),
			SymlinkPair({
				"source": "nwse-resize",
				"outcomes": [
					"nw-resize",
					"se-resize",
					"top_left_corner",
					"bottom_right_corner",
					"size_fdiag"
				]
			}),
			SymlinkPair({
				"source": "all-scroll",
				"outcomes": [
					"fleur",
					"cell",
					"move",
					"crosshair",
					"grabbing",
					"closedhand",
					"dnd-move",
					"cross",
					"plus",
					"size_all"
				]
			}),
			SymlinkPair({
				"source": "watch",
				"outcomes": [
					"wait",
					"progress",
					"half-busy",
					"left_ptr_watch"
				]
			})
		]
	})
	cursor_builder = CursorBuilder({ "cursor": dragon_byte })
	cursor_builder.build()

main()

